### ISSUE TITLE
Drop requirement to extend ObjectBehavior

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#1240]
Inheritance sucks, right?

We could provide a Trait that can be dropped into any class to make it a spec

(we could also auto-include the Trait using the stream wrapper, if that isn't a terrible idea)

**Comment by jakzal** [comment]
What problem(s) would this solve?

**Comment by ciaranmcnulty** [comment]
We'd be cooler

**Comment by ddziaduch** [comment]
> Inheritance sucks, right?
Why?

**Comment by drupol** [comment]
Prefer composition over it.

**Comment by Bilge** [comment]
phpspec will never be cool. It tries so hard to be cool it ends up lobotomizing itself in the process.

**Comment by ciaranmcnulty** [comment]
@Bilge not sure what this is adding to the conversation?

**Comment by ddziaduch** [comment]
@Bilge you should read https://inviqa.com/blog/my-top-ten-favourite-phpspec-limitations

**Comment by drupol** [comment]
> @Bilge you should read https://inviqa.com/blog/my-top-ten-favourite-phpspec-limitations

This article is from 2014, maybe an update would be nice, most probably that things have changed since then?

**Comment by ddziaduch** [comment]
@drupol it is quiet old but still valid :)

**Comment by Bilge** [comment]
Too bad it undermines itself with its first point.

**Comment by gquemener** [comment]
After a quick prototyping, this is quite feasible without too much refactoring/complexity.

I believe it's a matter of defining a trait with the same content as the `ObjectBehavior` class and a `SpecTransformer` that will adds the required `implements` statements to the spec class (when the trait usage is detected in the class).

It is possible to achieve this by keeping backward compatibility (by ignoring the spec transformation if the spec extends the `ObjectBehavior` class for instance).

Would this feature have a chance to gather a consensus among the maintainers?

**Comment by gquemener** [comment]
tl;dr;

is this a :+1: or :-1: idea?

```patch
  <?php

  namespace spec\App;

- use PhpSpec\ObjectBehavior;
+ use PhpSpec\Specifying;
  use App\MyOtherClass;

- class MyOtherClassSpec extends ObjectBehavior
  class MyOtherClassSpec
  {
+
+     use Specifying;
+
      function it_is_initializable()
      {
          $this->shouldHaveType(MyOtherClass::class);
      }
  }
```

**Comment by gquemener** [comment]
While going this way, why not going there ?

``` php
  <?php

  namespace spec\App;

- use PhpSpec\ObjectBehavior;
  use App\MyOtherClass;

- class MyOtherClassSpec extends ObjectBehavior
  spec MyOtherClassSpec
  {
      function it_is_initializable()
      {
          $this->shouldHaveType(MyOtherClass::class);
      }
  }
```

**Comment by gquemener** [comment]
Well, it would obviously break most IDE (syntax highlight/check), however nothing that config/plugin could not fix... eventually... :thinking:

**Comment by ddziaduch** [comment]
@gquemener the last one looks interesting, but how that would work?

**Comment by gquemener** [comment]
PhpSpec has the ability to run several [`SpecTransformer`](https://github.com/phpspec/phpspec/blob/master/src/PhpSpec/Loader/SpecTransformer.php) [prior to loading a spec class](https://github.com/phpspec/phpspec/blob/master/src/PhpSpec/Loader/StreamWrapper.php#L65-L67).

Creating a `SpecTransformer` that pre-process the new keyword into something php can parse is a possibility.

Take a look at this spec for a more in-depth example: https://github.com/phpspec/phpspec/blob/master/spec/PhpSpec/CodeAnalysis/TokenizedTypeHintRewriterSpec.php

**Comment by gquemener** [comment]
Over 9000 coolness level could be achieved with:


``` patch
  <?php

  namespace spec\App;

- use PhpSpec\ObjectBehavior;
  use App\MyOtherClass;

- class MyOtherClassSpec extends ObjectBehavior
+ describe MyOtherClass
  {
-     function it_is_initializable()
+     it is initializable()
      {
          $this->shouldHaveType(MyOtherClass::class);
      }
  }
```

**Comment by ddziaduch** [comment]
@gquemener so maybe just use functions like `describe()`, `it()` something like in Kahlan? 🙂

**Comment by gquemener** [comment]
yep indeed, that would almost achieve the same "readability" purpose without the pre-processing hassle.

I believe I would rather have these "helper" functions provided by PhpSpec rather than having some magical file pre-processing (although that it looks kind of cool!).

However, I may have missed the initial point and not understand how @ciaranmcnulty was intending to remove the parent class from specs, I don't see another solution than the spec transformer.
